src/_bentoml_impl/frameworks/unsloth/__init__.py

```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
+
+import logging
+import os
+import pathlib
+import shutil
+import subprocess
+import sys
+import tempfile
+import typing as t
+
+import yaml
+from deepmerge.merger import Merger
+
+import bentoml
+from _bentoml_sdk.service.config import ServiceConfig as Config
+from bentoml._internal.bento.build_config import BentoBuildConfig
+from bentoml._internal.bento.build_config import DockerOptions
+from bentoml._internal.bento.build_config import ModelSpec
+from bentoml._internal.utils import pkg
+from bentoml.exceptions import BentoMLException
+from bentoml.exceptions import MissingDependencyException
+
+from .mapping import RUNTIME_MAPPING as MAPPINGS
+from .mapping import get_extras
+
+logger = logging.getLogger(__name__)
+
+if t.TYPE_CHECKING:
+    from transformers import PreTrainedModel
+    from transformers import PreTrainedTokenizerFast
+
+if pkg.find_spec("unsloth") is None:
+    raise MissingDependencyException(
+        "'unsloth' is required in order to use module 'bentoml.unsloth', install unsloth with 'pip install bentoml[unsloth]'."
+    )
+
+merger = Merger(
+    # merge dicts, append list
+    [(dict, "merge"), (list, "append")],
+    # override all other types
+    ["override"],
+    # override conflicting types
+    ["override"],
+)
+
+
+def replace_tag(tag: str) -> str:
+    return tag.lower().replace("/", "--")
+
+
+ModelType = t.Literal["llama", "mistral", "gemma", "gemma2", "qwen2"]
+
+
+def build_bento(
+    model: PreTrainedModel,
+    tokenizer: PreTrainedTokenizerFast,
+    /,
+    model_name: str | None = None,
+    *,
+    quantization_method: t.Literal["bitsandbytes"] | None = None,
+    save_method: t.Literal["merged_16bit", "merged_4bit"] = "merged_16bit",
+    service_config: Config | None = None,
+    engine_config: dict[str, t.Any]
+    | None = None,  # arguments to pass to AsyncEngineArgs
+) -> bentoml.Model:
+    # this model is local then model_name must specified, otherwise derived from model_id
+    is_local = getattr(model.config, "_commit_hash", None) is None
+    if is_local is True and model_name is None:
+        raise BentoMLException(
+            'Fine-tune from a local checkpoint requires specifying "model_name".'
+        )
+    else:
+        model_name = model_name or replace_tag(model.config._name_or_path)
+
+    model_type = t.cast(ModelType, model.config.model_type)
+
+    if service_config is None:
+        service_config = {}
+    if engine_config is None:
+        engine_config = {}
+
+    service_config.update({**MAPPINGS[model_type]["service_config"]})
+
+    engine_config.update(MAPPINGS[model.config.model_type]["engine_config"])
+    if quantization_method is not None:
+        engine_config.update(
+            {"quantization": quantization_method, "load_format": quantization_method}
+        )
```

Comment:
Why do we take `quantization_method` out as a separate parameter instead of relying on `engine_config`?

Reply:
because for users who use bnb they also have to set load_format.

Reply:
I think they are better down through documentation. This abstraction is not clear about its override priority. If `quantization` and `load_format` must be the same, we can add engine_config validation logic.